File: src/state_machine.cpp
#include "state_machine.h"


StateMachine::
StateMachine() : current_state(SLEEP) {}


bool StateMachine::
change_state(const int32_t & new_state) {
    bool valid = false;

    /**
     * since this can be call in mutltiple thread need to lock mutex for the
     * whole function.
     */
    state_mutex.lock();

    switch(current_state) {
        case FAILED:
            if(new_state == FAILED)
                valid = true;
            break;

        case SLEEP:
            if(new_state == SLEEP)
                valid = true;
            break;

        case DETUMBLE:
            if(new_state == DETUMBLE || new_state == SLEEP || new_state == FAILED)
                valid = true;
            break;

        case POINT:
            if(new_state == POINT || new_state == SLEEP || new_state == FAILED)
                valid = true;
            break;

        default :
            cout << "state change failed" << endl; // TODO change to syslog later
    }

    if(valid) // transistion was valid, change the state.
        current_state = new_state;

    state_mutex.unlock();
    return valid;
}


int32_t StateMachine::
get_current_state() {
    return current_state;
}

File: adcs_manager/state_machine.py
from enum import Enum
from threading import Lock



class State(Enum):
    FAILED = 0      # Something failed.
    SLEEP = 1       # ADCS program is not doing anything and waiting for commands.
    DETUMBLE = 2    # Detumble the satellite after deployement
    POINT = 3       # Point the satellite for CFC / Oresat Live


class StateMachine:
    def __init__(self):
        self._lock = Lock()
        self._current_state = State.SLEEP.value
        self._valid_transition = {
                State.FAILED.value : [State.FAILED.value, State.SLEEP.value],
                State.SLEEP.value : [State.SLEEP.value, State.DETUMBLE.value, State.POINT.value],
                State.DETUMBLE.value : [State.DETUMBLE.value, State.SLEEP.value, State.FAILED.value],
                State.POINT.value : [State.POINT.value, State.SLEEP.value, State.FAILED.value],
            }


    def change_state(self, new_state):
        # (int) -> bool
        """
        Check if the state change is valid.
        Returns true if valid and transition happend
        Returns false if errored.
        """

        self._lock.acquire()

        # check if new_state is a valid transition
        if new_state in self._valid_transition[self._current_state]:
            self._current_state = new_state
            valid = True
        else:
            self._current_state = State.FAILED.value
            valid = False # not a valid transition

        self._lock.release()
        return valid


    def get_current_state(self):
        return self._current_state
